src/solvers/refinement/string_refinement.cpp

```diff
@@ -1772,9 +1777,165 @@ bool string_refinementt::is_axiom_sat(
+  bool correct;
+
+  is_linear_visitort(): correct(true) {}
+
+  void operator()(const exprt &expr) override
+  {
+    if(expr.id()!=ID_plus && expr.id()!=ID_minus && expr.id()!=ID_unary_minus)
+    {
+      // This represents that the expr is a valid leaf, may not be future proof
+      // or 100% enforced, but is correct prescriptively. All non-sum exprs must
+      // be leaves.
+      correct&=expr.operands().empty();
+    }
+  }
+};
+
+/// \related string_constraintt
+static bool is_linear(const exprt &expr)
+{
+  is_linear_visitort v;
+  expr.visit(v);
+  return v.correct;
+}
+
+/// \related string_constraintt
+static bool universal_only_in_index(const string_constraintt &expr)
+{
+  // We need to check that each instance of the universal variable occurs in
+  // and index expression. For efficiency, we do a depth-first search of the
+  // body. The exprt visitors do a BFS and hide the stack/queue, so we would
+  // need to store a map from child to parent.
+
+  // The int represents index depth we are.
+  typedef std::pair<exprt, unsigned> valuet;
+  std::stack<valuet> stack;
+  // How many index expressions we are currently in. We start at 0 since expr
+  // is not an index expression.
+  stack.push(valuet(expr.body(), 0));
+  while(!stack.empty())
+  {
+    // Inspect current value
+    const valuet cur=stack.top();
+    stack.pop();
+    const exprt e=cur.first;
+    const unsigned index_depth=cur.second;
+    const unsigned child_index_depth=index_depth+(e.id()==ID_index?0:1);
+
+    // If we found the universal variable not in an index_exprt, fail
+    if(e==expr.univ_var() && index_depth==0)
+      return false;
+    else
+      forall_operands(it, e)
+        stack.push(valuet(*it, child_index_depth));
+  }
+  return true;
+}
+
+/// Checks the data invariant for \link string_constraintt
+/// \related string_constraintt
+/// \param [in] expr: the string constraint to check
+/// \return whether the constraint satisfies the invariant
+bool string_refinementt::is_valid_string_constraint(
+  const string_constraintt &expr)
+{
+  // Condition 1: The premise cannot contain any string indices
+  const array_index_mapt premise_indices=gather_indices(expr.premise());
+  if(!premise_indices.empty())
+  {
+    error() << "Premise has indices: " << from_expr(ns, "", expr) << ", map: {";
+    for(const auto &pair : premise_indices)
+    {
+      error() << from_expr(ns, "", pair.first) << ": {";
+      for(const auto &i : pair.second)
+        error() << from_expr(ns, "", i) <<  ", ";
+    }
+    error() << "}}" << eom;
+    return false;
+  }
+
+  const array_index_mapt body_indices=gather_indices(expr.body());
+  // Must validate for each string. Note that we have an invariant that the
+  // second value in the pair is non-empty.
+  for(const auto &pair : body_indices)
+  {
+    // Condition 2: All indices of the same string must be the of the same form
+    const exprt rep=pair.second.back();
+    for(size_t j=0; j<pair.second.size()-1; j++)
```

Comment:
If there's a way to use the ranged for for all elements except the first or last, then yes. Otherwise, I did this to avoid comparing an element with itself.